File: drift/crypto/sealed.py
"""
drift.crypto.sealed — sealed sender (Phase 3b)

Removes the last sender-correlatable metadata from the clear envelope. Before
this, the relay saw the sender's ephemeral key ``R`` and — more importantly —
the Double Ratchet header, whose DH ratchet public key stays constant across a
sender's messages within a ratchet epoch and so *links* them. Tor hid the
sender's IP, but the relay could still group a sender's traffic by that header.

Sealed sender folds everything except the recipient's one-time detection
address into a single opaque blob. On the wire the relay now sees only:

  - ``addr``  the recipient's one-time stealth address  (routing/detection)
  - ``ct``    one opaque blob

The blob is::

    R (32 bytes)  ||  u16(len)  ||  sealed_header  ||  ratchet_ciphertext

``R`` is the per-message ephemeral public key. It is *not* encrypted, and it
cannot be: it is the Diffie-Hellman contribution the recipient needs to derive
the stealth secret in the first place (encrypting it under a key derived from
itself is circular). This is the one unavoidable clear value in any ECDH
stealth scheme — but it is fresh every message and reveals nothing about the
sender's identity (DRIFT carries no sender identity to begin with). What this
module *does* seal is the ratchet header, the field that actually linked a
sender's messages.

The sealing key is derived from the stealth shared secret that
``derive_one_time_address`` / ``scan_for_message`` already compute (and which,
until now, the session discarded). Both sides reach the same key:

    seal_key = HKDF(stealth_key, info="drift-sealed-sender-v1")

The recipient's one-time address is bound in as AEAD associated data, so the
relay cannot move a sealed blob onto a different address without the unseal
failing.

This module composes existing primitives (``drift.crypto.encrypt`` /
``derive_message_key``); it implements no curve or cipher math itself.
"""

from __future__ import annotations

import struct

from drift.crypto import decrypt, derive_message_key, encrypt

# Domain separation for the header-sealing key. Versioned: changing it breaks
# wire compatibility with already-deployed clients.
SEAL_INFO = b"drift-sealed-sender-v1"

# Ephemeral X25519 public keys are 32 bytes; the sealed-header length prefix is
# a 2-byte big-endian integer (the ratchet header is well under 64 KiB).
_EPK_LEN = 32
_LEN_PREFIX = struct.Struct(">H")


def _seal_key(stealth_key: bytes) -> bytes:
    """Derive the header-sealing key from the per-message stealth secret."""
    return derive_message_key(stealth_key, info=SEAL_INFO)


def seal(
    stealth_key: bytes,
    ephemeral_pub: bytes,
    ratchet_header: bytes,
    ratchet_ciphertext: bytes,
    *,
    address: bytes = b"",
) -> bytes:
    """
    Build the opaque sealed-sender blob.

    ``stealth_key`` is the per-message key returned by
    ``derive_one_time_address``. ``ephemeral_pub`` is R. ``ratchet_header`` is
    the serialized Double Ratchet header to be sealed. ``address`` is the
    recipient one-time address, bound in as associated data.
    """
    if len(ephemeral_pub) != _EPK_LEN:
        raise ValueError(f"ephemeral_pub must be {_EPK_LEN} bytes, got {len(ephemeral_pub)}")
    sealed_header = encrypt(_seal_key(stealth_key), ratchet_header, associated_data=address)
    return b"".join((
        ephemeral_pub,
        _LEN_PREFIX.pack(len(sealed_header)),
        sealed_header,
        ratchet_ciphertext,
    ))


def parse(blob: bytes) -> tuple[bytes, bytes, bytes]:
    """
    Split a blob into ``(ephemeral_pub, sealed_header, ratchet_ciphertext)``.

    Pure framing — no decryption, no key needed (the recipient needs the
    ephemeral key out before it can derive anything). Raises ``ValueError`` if
    the blob is too short or internally inconsistent, which the caller treats as
    "not a well-formed stealth message" and skips.
    """
    if len(blob) < _EPK_LEN + _LEN_PREFIX.size:
        raise ValueError("sealed blob too short for header")
    ephemeral_pub = blob[:_EPK_LEN]
    (slen,) = _LEN_PREFIX.unpack_from(blob, _EPK_LEN)
    start = _EPK_LEN + _LEN_PREFIX.size
    end = start + slen
    if end > len(blob):
        raise ValueError("sealed blob truncated")
    sealed_header = blob[start:end]
    ratchet_ciphertext = blob[end:]
    return ephemeral_pub, sealed_header, ratchet_ciphertext


def open_header(stealth_key: bytes, sealed_header: bytes, *, address: bytes = b"") -> bytes:
    """
    Decrypt a sealed ratchet header.

    Only called after the stealth scan has already confirmed the message is
    ours, so an authentication failure here is genuine tampering — ``InvalidTag``
    is allowed to propagate (the iron rule), never swallowed.
    """
    return decrypt(_seal_key(stealth_key), sealed_header, associated_data=address)

File: drift-core/crypto/src/sealed.rs
//! Sealed sender, matching `drift.crypto.sealed`.
//!
//! Folds everything except the recipient's one-time address into one opaque
//! blob: `R(32) || u16(len) || sealed_header || ratchet_ciphertext`. The header
//! is sealed under `HKDF(stealth_key, "drift-sealed-sender-v1")` with the
//! recipient one-time address bound as AEAD associated data.

use zeroize::Zeroize;

use crate::aead::{decrypt, encrypt_with_nonce};
use crate::kdf::derive_message_key;
use crate::{Error, Result};

const SEAL_INFO: &[u8] = b"drift-sealed-sender-v1";
const EPK_LEN: usize = 32;

fn seal_key(stealth_key: &[u8]) -> [u8; 32] {
    derive_message_key(stealth_key, None, SEAL_INFO)
}

/// Build the opaque sealed-sender blob. `nonce` is supplied so a vector can be
/// reproduced; production draws it from the OS CSPRNG (see [`seal`]).
pub fn seal_with_nonce(
    stealth_key: &[u8],
    nonce: &[u8; 24],
    ephemeral_pub: &[u8; 32],
    ratchet_header: &[u8],
    ratchet_ciphertext: &[u8],
    address: &[u8],
) -> Vec<u8> {
    let mut key = seal_key(stealth_key);
    let sealed_header = encrypt_with_nonce(&key, nonce, ratchet_header, address);
    key.zeroize();
    let mut out = Vec::with_capacity(EPK_LEN + 2 + sealed_header.len() + ratchet_ciphertext.len());
    out.extend_from_slice(ephemeral_pub);
    out.extend_from_slice(&(sealed_header.len() as u16).to_be_bytes());
    out.extend_from_slice(&sealed_header);
    out.extend_from_slice(ratchet_ciphertext);
    out
}

/// Build a sealed blob with a fresh random AEAD nonce.
pub fn seal(
    stealth_key: &[u8],
    ephemeral_pub: &[u8; 32],
    ratchet_header: &[u8],
    ratchet_ciphertext: &[u8],
    address: &[u8],
) -> Vec<u8> {
    use rand_core::{OsRng, RngCore};
    let mut nonce = [0u8; 24];
    OsRng.fill_bytes(&mut nonce);
    seal_with_nonce(
        stealth_key,
        &nonce,
        ephemeral_pub,
        ratchet_header,
        ratchet_ciphertext,
        address,
    )
}

/// Split a blob into `(ephemeral_pub, sealed_header, ratchet_ciphertext)`. Pure
/// framing — no key needed (the recipient needs R out before deriving anything).
pub fn parse(blob: &[u8]) -> Result<([u8; 32], &[u8], &[u8])> {
    if blob.len() < EPK_LEN + 2 {
        return Err(Error::Malformed("sealed blob too short for header"));
    }
    let mut ephemeral_pub = [0u8; 32];
    ephemeral_pub.copy_from_slice(&blob[..EPK_LEN]);
    let slen = u16::from_be_bytes([blob[EPK_LEN], blob[EPK_LEN + 1]]) as usize;
    let start = EPK_LEN + 2;
    let end = start + slen;
    if end > blob.len() {
        return Err(Error::Malformed("sealed blob truncated"));
    }
    Ok((ephemeral_pub, &blob[start..end], &blob[end..]))
}

/// Decrypt a sealed ratchet header. `Error::InvalidTag` on tampering.
pub fn open_header(stealth_key: &[u8], sealed_header: &[u8], address: &[u8]) -> Result<Vec<u8>> {
    let mut key = seal_key(stealth_key);
    let out = decrypt(&key, sealed_header, address);
    key.zeroize();
    out
}
